pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,371 @@
+			if o.activeSerie.Load() > 0 {
+				out <- prometheus.MustNewConstMetric(t.activeSeriesPerUserAttribution, prometheus.GaugeValue, o.activeSerie.Load(), keys...)
+			}
+			if o.receivedSample.Load() > 0 {
+				out <- prometheus.MustNewConstMetric(t.receivedSamplesAttribution, prometheus.CounterValue, o.receivedSample.Load(), keys...)
+			}
+			o.discardedSampleMtx.Lock()
+			for reason, discarded := range o.discardedSample {
+				out <- prometheus.MustNewConstMetric(t.discardedSampleAttribution, prometheus.CounterValue, discarded.Load(), append(keys, reason)...)
+			}
+			o.discardedSampleMtx.Unlock()
+		}
+	}
+	if t.totalFailedActiveSeries.Load() > 0 {
+		out <- prometheus.MustNewConstMetric(t.failedActiveSeriesDecrement, prometheus.CounterValue, t.totalFailedActiveSeries.Load(), t.userID)
+	}
+}
+
+func (t *Tracker) IncrementDiscardedSamples(lbs []mimirpb.LabelAdapter, value float64, reason string, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now.Unix(), 0, 0, value, &reason, true)
+}
+
+func (t *Tracker) IncrementReceivedSamples(lbs []mimirpb.LabelAdapter, value float64, now time.Time) {
+	if t == nil {
+		return
+	}
+	t.updateCountersWithLabelAdapter(lbs, now.Unix(), 0, value, 0, nil, true)
+}
+
+func (t *Tracker) IncrementActiveSeriesFailure() {
+	if t == nil {
+		return
+	}
+	t.totalFailedActiveSeries.Add(1)
+}
+
+func (t *Tracker) updateCountersWithLabelAdapter(lbls []mimirpb.LabelAdapter, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	extractValues := func() []string {
+		labelValues := make([]string, len(t.labels))
+		for idx, cal := range t.labels {
+			for _, l := range lbls {
+				if l.Name == cal {
+					labelValues[idx] = l.Value
+					break
+				}
+			}
+			if labelValues[idx] == "" {
+				labelValues[idx] = missingValue
+			}
+		}
+		return labelValues
+	}
+	t.updateCountersCommon(extractValues, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+}
+
+func (t *Tracker) updateCounters(lbls labels.Labels, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string, createIfDoesNotExist bool) {
+	extractValues := func() []string {
+		labelValues := make([]string, len(t.labels))
+		for idx, cal := range t.labels {
+			labelValues[idx] = lbls.Get(cal)
+			if labelValues[idx] == "" {
+				labelValues[idx] = missingValue
+			}
+		}
+		return labelValues
+	}
+	t.updateCountersCommon(extractValues, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason, createIfDoesNotExist)
+}
+
+func (t *Tracker) updateCountersCommon(
+	extractValues func() []string,
+	ts int64,
+	activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64,
+	reason *string,
+	createIfDoesNotExist bool,
+) {
+	// Extract label values
+	labelValues := extractValues()
```

Comment:
Why are we passing down a callback here instead of just passing the `labelValues []string` to this method?

Reply:
addressed in 7a60c7d7c63f5a8dd5385e2a0af8e3644b46b529, because I thought i was smart